torch/nn/modules/layernorm.py

```diff
@@ -0,0 +1,116 @@
+from .batchnorm import _BatchNorm
+from .. import functional as F
+
+
+class LayerNorm(_BatchNorm):
+    r"""Applies Layer Normalization over a mini-batch of inputs as described in
+    the paper `Layer Normalization`_ .
+
+    .. math::
+        y = \frac{x - mean[x]}{ \sqrt{Var[x]} + \epsilon} * gamma + beta
+
+    The mean and standard-deviation are calculated separately for all dimensions
```

Comment:
This is a little unclear. What layernorm should do, as far as I know, is compute means and stdevs over a single dimension (representing the channel, typically dimension -1) or occasionally in the case of image data multiple dimensions (i.e. one or more spatial dims + the channel dim). But at least for the NLP use case it shouldn't reduce over all non-batch dimensions, only one of them, and it should never compute statistics over the batch dimension.

Reply:
From the way the original paper describes the concept and formulates LN, LN normalizes all inputs to a layer. So it's across dimensions. tf.contrib implementation also agrees on this https://www.tensorflow.org/api_docs/python/tf/contrib/layers/layer_norm. That said, it is not quite clear what LN does when input has channels/planes (e.g. images) since it makes more sense to have it behave like IN. @Kaixhin and I had a discussion on this, and we hoped to get a concrete answer from the author. Unfortunately, we haven't heard back.

Reply:
And yeah, it doesn't compute statistics across the batch dimension. The implementation doesn't do so. I need to fix the doc :)

Reply:
The TF implementation has a parameterizable number of dimensions to normalize over; it always normalizes over the last `n` dimensions but `n` is configurable. That seems like an OK API, since the channel dimension is virtually always last.
The original paper normalizes over "all the hidden units in a layer," which is a little ambiguous, but clarifies that in recurrent networks the normalization is performed "at the current timestep," which would translate to normalizing only over the channel dimension if the input tensor also had a time dimension; that's the strategy adopted in e.g. [the Transformer network](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/common_layers.py#L517).

Reply:
Your argument on RNN is valid. The problem with RNN is that it has two aggregating dimensions (batch and time). It's my bad for forgetting LN's popularity in NLP and focusing only on use cases similar to BN and IN. I believe an API like the tf.contrib one is reasonable in this case. 

However, what is the difference between IN and LN then? From the user's perspective, what LN can do is a superset of what IN can do.

Reply:
In TF the order for images is `NHWC` but for us it's `NCHW` (same type of ordering for 1D and 3D signals), so we can't copy their API directly. For RNNs which take an entire sequence, we have `TNC` by default, or `NTC` if `batch_first = True`. I may have got this wrong, but LN/IN (which I'm considering the same atm, with different use-cases explained in the docs) should choose the first dim (from the left) to normalise, and then collapse everything to the right into this dim?

Reply:
@Kaixhin Yes, that is what I'm thinking about as well. (It is essentially the tf API btw.)

Reply:
Great so we have something like `begin_norm_axis` but can get rid of `begin_params_axis` because of our ordering?

Reply:
Yes. I plan to only have `begin_norm_axis` (maybe with a different name).